File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.py
import sys
from collections import deque


def bfs(graph, queue):
  dx = [2, 2, 1, 1, -1, -1, -2, -2]
  dy = [1, -1, 2, -2, 2, -2, 1, -1]
  while queue:
    cur = queue.popleft()
    for i in range(8):
      nxt_x = cur[0] + dx[i]
      nxt_y = cur[1] + dy[i]
      if 0 <= nxt_x < len(graph) and 0 <= nxt_y < len(graph) and graph[nxt_x][nxt_y] == 0:
        graph[nxt_x][nxt_y] = graph[cur[0]][cur[1]] + 1
        queue.append((nxt_x, nxt_y))

        
    

def solution():
  answer = []
  N = int(sys.stdin.readline())
  for _ in range(N):
    i = int(sys.stdin.readline())
    graph = [[0 for _ in range(i)] for _ in range(i)]
    x, y = map(int, sys.stdin.readline().split())
    tar_x, tar_y = map(int, sys.stdin.readline().split())
    if x == tar_x and y == tar_y:
      answer.append(0)
      continue 
    queue = deque([(x, y)])
    bfs(graph, queue)
    answer.append(graph[tar_x][tar_y])
  for element in answer:
    print(element)
    
    
solution()

File: 백준/Silver/7562. 나이트의 이동/나이트의 이동.js
let fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const dx = [-2, -1, 1, 2, 2, 1, -1, -2];
const dy = [1, 2, 2, 1, -1, -2, -2, -1];

const bfs = (queue, I, visited, target) => {
    const [tarX, tarY] = target;
    while (queue.length > 0) {
        const [curX, curY, depth] = queue.shift();
        if(curX === tarX && curY === tarY) return depth;
        for (let i = 0; i < 8; i++) {
            const nextX = curX + dx[i];
            const nextY = curY + dy[i];
            if (nextX < 0 || nextY < 0 || nextX >= I || nextY >= I) continue;
            if (visited[nextX][nextY] !== 1) {
                queue.push([nextX, nextY, depth + 1]);
                visited[nextX][nextY] = 1;
            }
        }
    }
};

const answer = [];

for (let i = 1; i < input.length; i++) {
    const I = Number(input[i]);
    const queue = [];
    const visited = Array.from({length: I}, () => Array.from({length: I}, () => 0));
    i++;
    const [curX, curY] = input[i].split(' ').map(Number);
    visited[curX][curY] = 1;
    queue.push([curX, curY, 0]);
    i++;
    const [tarX, tarY] = input[i].split(' ').map(Number);
    answer.push(bfs(queue, I, visited, [tarX, tarY]));
}

let answerString = '';

for (let i = 0; i < answer.length; i++) {
    answerString = answerString + answer[i] + '\n';
}

console.log(answerString.trim());